### PR TITLE
Limiting cold storage of sliding sync timeline items, fix bugs

### PR DESCRIPTION
**Pull request by gnunicorn**

### CODE REVIEW

File: crates/matrix-sdk/src/sliding_sync.rs [hunk 1]
```diff
@@ -162,11 +156,19 @@ struct FrozenSlidingSyncRoom {
...
 impl From<&SlidingSyncRoom> for FrozenSlidingSyncRoom {
     fn from(value: &SlidingSyncRoom) -> Self {
+        let locked_tl = value.timeline.lock_ref();
+        let tl_len = locked_tl.len();
+        let (prev_batch, timeline) = if tl_len > 10 {
+            let pos = tl_len - 10;
+            (None, locked_tl.iter().skip(pos).cloned().collect())
+        } else {
+            (value.prev_batch.lock_ref().clone(), locked_tl.to_vec())
+        };
```
Comment (jplatte) [inline crates/matrix-sdk/src/sliding_sync.rs:166]: A comment that explains why this is done would be useful I think.

File: crates/matrix-sdk/src/sliding_sync.rs [hunk 2]
```diff
@@ -1406,6 +1410,11 @@ impl SlidingSyncView {
             changed = true;
         }
 
+        if self.rooms_count.get() != Some(rooms_count) {
+            self.rooms_count.replace(Some(rooms_count));
+            changed = true
```
Comment (jplatte) [inline crates/matrix-sdk/src/sliding_sync.rs:1415]:
```suggestion
            changed = true;
```

File: crates/matrix-sdk/src/sliding_sync.rs [hunk 3]
```diff
@@ -1406,6 +1410,11 @@ impl SlidingSyncView {
             changed = true;
         }
 
+        if self.rooms_count.get() != Some(rooms_count) {
+            self.rooms_count.replace(Some(rooms_count));
```
Comment (jplatte) [inline crates/matrix-sdk/src/sliding_sync.rs:1414]:
```suggestion
            self.rooms_count.set(Some(rooms_count));
```

**Review by jplatte — APPROVED** [review]

File: crates/matrix-sdk/src/sliding_sync.rs (hunk 2 above)
Comment (gnunicorn) [reply crates/matrix-sdk/src/sliding_sync.rs:1415]: why isn't this something rustfmt picks up? same for missing `;` on return ...

Comment (jplatte) [reply crates/matrix-sdk/src/sliding_sync.rs:1415]: I opened a feature request for it a while ago, maybe the new rustfmt team will implement it next year 🤷🏼 https://github.com/rust-lang/rustfmt/issues/5363

**Pull request merged**